src/lib.rs

```diff
@@ -97,9 +98,61 @@ mod accept {
     }
 }
 
+mod host {
+    /// The host part of a domain (without scheme, port and path).
+    ///
+    /// This is the argument to the [`connect`](crate::connect::connect) function. Strings and string slices are
+    /// converted into Hosts automatically, as is [Url](url::Url) with the `host-from-url` feature (enabled by default).
+    #[derive(Debug)]
+    pub struct Host(String);
+
+    impl Host {
+        /// The host as string. Consumes self.
+        pub fn as_string(self) -> String {
+            self.0
+        }
+    }
+
+    impl From<&str> for Host {
+        fn from(host: &str) -> Self {
+            Self(host.into())
+        }
+    }
+
+    impl From<String> for Host {
+        fn from(host: String) -> Self {
+            Self(host)
+        }
+    }
+
+    #[cfg(feature = "host-from-url")]
+    use url::Url;
+    #[cfg(feature = "host-from-url")]
+    impl From<Url> for Host {
+        fn from(url: Url) -> Self {
+            Self(
+                url.host_str()
+                    .expect("URL has to include a host part.")
+                    .into(),
+            )
+        }
+    }
+    #[cfg(feature = "host-from-url")]
+    impl From<&Url> for Host {
+        fn from(url: &Url) -> Self {
+            Self(
+                url.host_str()
+                    .expect("URL has to include a host part.")
+                    .into(),
+            )
+        }
+    }
+}
+
```

Comment:
Rather than re-declaring the #[`cfg(feature)]` blocks repeatedly, it can be applied to blocks directly. See https://github.com/async-rs/async-std/blob/master/src/utils.rs#L83-L105 for the generated macro code we use in `async-std` where we do a similar thing.